packages/artifact/src/internal-artifact-client.ts

```diff
@@ -100,25 +133,114 @@ export class DefaultArtifactClient implements ArtifactClient {
     return uploadResponse
   }
 
-  /*
-  Downloads a single artifact associated with a run
+  async downloadArtifact(
+    name: string,
+    path?: string | undefined,
+    options?: DownloadOptions | undefined
+  ): Promise<DownloadResponse> {
+    const artifacts = await listArtifacts()
+    if (artifacts.count === 0) {
+      throw new Error(
+        `Unable to find any artifacts for the associated workflow`
+      )
+    }
+
+    const artifactToDownload = artifacts.value.find(artifact => {
+      return artifact.name === name
+    })
+
+    if (!artifactToDownload) {
+      throw new Error(`Unable to find an artifact with the name: ${name}`)
+    } else {
+      // query container for all items associated with the artifact
+      const items = await getContainerItemsForSingleArtifact(
+        artifactToDownload.name,
+        artifactToDownload.fileContainerResourceUrl
+      )
+
+      if (!path) {
+        path = getWorkSpaceDirectory()
+      }
 
-  export async function downloadArtifact(
-      name: string,
-      path?: string,
-      options?: DownloadOptions
-    ): Promise<DownloadResponse> {
+      // Check only for downloadable items that are not directories. During upload, empty directories are rejected by the remote server so there should be no
+      // artifacts that consist of only empty directories
+      if (items.value.length === 0) {
+        throw new Error(
+          `No downloadable files were found for the artifact: ${artifactToDownload.name}`
+        )
+      } else {
+        const downloadSpecification = getDownloadSpecification(
+          name,
+          items.value,
+          path,
+          options?.createArtifactFolder || false
+        )
+        await createDirectoriesForArtifact(
+          downloadSpecification.directoryStructure
+        )
+        await downloadSingleArtifact(downloadSpecification.filesToDownload)
 
-    TODO
+        return {
+          artifactName: name,
+          downloadPath: downloadSpecification.rootDownloadLocation
+        }
+      }
+    }
   }
 
-  Downloads all artifacts associated with a run. Because there are multiple artifacts being downloaded, a folder will be created for each one in the specified or default directory
+  async downloadAllArtifacts(
+    path?: string | undefined
+  ): Promise<DownloadResponse[]> {
+    const response: DownloadResponse[] = []
+    const artifacts = await listArtifacts()
+    if (artifacts.count === 0) {
+      throw new Error(
```

Comment:
Should not throw here, just log an info message.  We throw if `listArtifacts` throw, otherwise this is perfectly fine.  The action itself can decide if they want to throw or not.

Reply:
Updated. Just `core.info` is used now and an empty list is returned